patches/sqlite/0002-sqlite-complete-early-exit.patch

```diff
@@ -0,0 +1,64 @@
+--- sqlite-src-pristine/src/complete.c	2023-11-01 22:31:37
++++ sqlite-src-modified/src/complete.c	2024-03-12 12:48:48
+@@ -99,14 +99,21 @@
+ **
+ ** If we compile with SQLITE_OMIT_TRIGGER, all of the computation needed
+ ** to recognize the end of a trigger can be omitted.  All we have to do
+ ** is look for a semicolon that is not part of an string or comment.
+ */
+-int sqlite3_complete(const char *zSql){
++int sqlite3_complete(const char *zSql) {
++    return sqlite3_complete_length(zSql, 0) != 0;
++}
++
++int sqlite3_complete_length(const char *zSql, int firstOnly){
+   u8 state = 0;   /* Current state, using numbers defined in header comment */
+   u8 token;       /* Value of the next token */
+
++  /* Store the original pointer for later index calculations */
++  const char *initialZSql = zSql;
++
+ #ifndef SQLITE_OMIT_TRIGGER
+   /* A complex statement machine used to detect the end of a CREATE TRIGGER
+   ** statement.  This is the normal case.
+   */
+   static const u8 trans[8][8] = {
+@@ -252,14 +259,18 @@
+           token = tkOTHER;
+         }
+         break;
+       }
+     }
+-    state = trans[state][token];
+     zSql++;
++    state = trans[state][token];
++
++    // Once we have found a single complete statement, return its length
++    if (firstOnly && state==1) return zSql - initialZSql;
+   }
+-  return state==1;
++  // Otherwise only return the length if the entire input is valid
++  return state==1 && zSql - initialZSql;
```

Comment:
In C/C++, an `&&` expression can evaluate only to 0 or 1. It does not evaluate to the right-side value when the left side is truthy.

Is there a missing test which should have caught this?

Reply:
That's weird, it's passing my tests. If this was returning `1` it would definitely be breaking them. But I changed it to this form because of build errors in MacOS, so I've switched it back to explicitly return `zSql - initialZSql` or `0`.